File: app/static/js/undo.js
function undo() {
    $.getJSON( $SCRIPT_ROOT + '/_undo').success( function(data) {
        action=data.result;
        undo_trash(action);
        if ( (action['act'] === "drive") || (action['act'] === "shuttle") || (action['act'] === "dispatch") ) {
            undo_move(action);$('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if ((action['act'] === "charter") ||
                   (action['act'] === "fly") ||
                   (action['act'] === "station-fly")) {
            undo_discard(action['cards'], action, action['owner']);
            undo_move(action);
            $('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if (action['act'] === "airlift") {
            undo_discard(action['cards'], action, action['owner']);
            undo_move(action);
        } else if (action['act'] === "build") {
            undo_station(action);
            $('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if (action['act'] === "gg") {
            undo_station(action);
        } else if (action['act'] === "treat") {
            undo_treatment(action);
            $('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if (action['act'] === "cure") {
            undo_cure(action);
            $('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if (action['act'] === "take") {
            undo_take(action);
            $('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if (action['act'] === "give") {
            undo_give(action);
            $('#actions-'+String(PHASE)).attr('class', 'off');
            PHASE--;
        } else if (action['act'] === "rp") {
            undo_rp(action);
        } else if (action['act'] === "oqn") {
            undo_oqn(action);
        } else if (action['act'] === "store") {
            undo_store(action);
        }
        ACTIONS--;
        if (PHASE < 5) {
            set_cities(data.available);
            set_giveable(data.hand, data.can_give);
            set_takeable(data.team_hands, data.can_take);
            $('#build-station').prop('disabled', !data.can_build);
            $('#make-cure').prop('disabled', !data.can_cure);
        }
        if (ACTIVE === 'cp') {
            set_cp_store();
        }
        events_on();
        $('#undo-action').prop('disabled', ACTIONS === 0);
        set_next_button();
    }).error(function(error){console.log(error)});
}

function undo_discard(card, action, owner) {
    if (action.is_stored) {
        $('#cp-store').off().on('click', select_store).show(200).attr('class', '.store-'+String(card)).addClass('giveable').attr('src', 'static/img/player_cards/pl-'+String(card)+'.svg');
    } else if ($('.active-player').attr('id').split('-')[1] !== owner) {
        $('#'+owner+"-card-"+card).show(200);
    } else {
        $('#card-'+card).show(200);
    }
    $("#pl-discard-"+card).hide(200);
    $('#logger').html( $('#logger').html()+' Returned '+CARDS[Number(card)].bold()+' to the '+ROLES[owner].bold()+"'s hand.");
}

function undo_move(action) {
    var map = Snap.select('#cities');
    var city = document.getElementById('city-'+String(action['origin']));
    var city_dims = city.getBoundingClientRect();
    var city_w = city_dims.width;
    var city_h = city_dims.height;
    var city_l = city_dims.left;
    var city_t = city_dims.top;
    var index = $('#'+action['mover']).parent().parent().index();
    var menu_on = 0;
    if ( $('body').hasClass('menu-push-toright') ) {
        menu_on = -1;
    } else if ( $('body').hasClass('menu-push-toleft') ) {
        menu_on = 1;
    }
    var player_t = city_t - 0.57*city_h;
    var player_l = city_l + (index+1)*3*city_w/8;
    for (var i=0; i<action['cubes'].length; i++) {
        if ($(".city-"+String(action['destination'])+".cube-"+String(i)).length !== action['cubes'][i]) {
            var cube_counter = document.getElementById(String(i)+"-cnt").getElementsByTagName('tspan')[0].textContent;
            var cubes_left = Number(cube_counter);
            var prev_city = document.getElementById('city-'+String(action['destination']));
            var prev_dims = prev_city.getBoundingClientRect();
            add_cubes(String(action['destination']), prev_dims, String(i), action['rows'][i], action['cubes'][i]);
            stop_svg(action['origin']);
            if ( $.inArray(Number(action['origin']), action['at_risk']) !== -1 ) {
                pulse_svg(action['origin']);
            }
            cubes_left += action['cubes'][i];
            document.getElementById(String(i)+"-cnt").getElementsByTagName('tspan')[0].textContent = String(cubes_left);
        }
    }

    map.selectAll(".available").forEach( function(el) {
		el.unavailable();
	});
    $("#"+action['mover']+"-piece").offset({left: player_l, top: player_t});
    set_cities(action['available']);
    if ( action['mover'] === ACTIVE ) {
        set_treatable(action['origin'])
    }
    set_giveable(action['hand'], action['give']);
    set_takeable(action['team_hands'], action['take']);
    $('#logger').html("(UNDO) Return "+ROLES[action['mover']].bold()+" to "+CARDS[action['origin']].bold()+".");
    $("#build-station").prop('disabled', !action['can_build']);
    $("#make-cure").prop('disabled', !action['can_cure']);
}

function undo_station(action) {
    var city = Number(action['origin']);
    $('#station-'+action['origin']).attr('class', 'unbuilt').hide();
    $('#logger').html('(UNDO) Removed station from '+CARDS[city].bold()+'.');
    if ( action['removed'] !== '-1' ) {
        $('#station-'+action['removed']).attr('class', 'built').show();
        $('#logger').html($('#logger').html()+' Returned station to '+CARDS[Number(action['removed'])].bold()+'.')
    }
    if ( action['cards'] !== -1) {
        undo_discard(action['cards'], action, action['owner']);
    }
    set_cities(action['available']);
    if (action['cards'] !== '50' || action['can_build']) {
        $("#build-station").prop('disabled', false);
    }
    if ( action['removed'] === 'none' ) {
        var stations_left = Number(document.getElementById("research-cnt").getElementsByTagName('tspan')[0].textContent)

        if (stations_left < 6)
        document.getElementById("research-cnt").getElementsByTagName('tspan')[0].textContent = stations_left + 1;
    }
    events_on();
}

function undo_treatment(action) {
    var city = document.getElementById('city-'+action['origin']);
    var dims = city.getBoundingClientRect();
    add_cubes(action['origin'], dims, action['color'],
                action['rows'][Number(action['color'])], action['removed'], action['at_risk']);
    stop_svg(action['origin']);
    if ( $.inArray(Number(action['origin']), action['at_risk']) !== -1 ) {
        pulse_svg(action['origin']);
    }

    var cubes_left = Number(document.getElementById(action['color']+"-cnt").getElementsByTagName('tspan')[0].textContent)
    document.getElementById(action['color']+"-cnt").getElementsByTagName('tspan')[0].textContent = cubes_left - action['removed'];
    set_treatable(action['origin']);
    $('#logger').html('(UNDO) Returned '+String(action['removed'])+' '+COLORS[Number(action['color'])].bold()+' cube(s) to '+CARDS[Number(action['origin'])].bold()+'.');
}

function undo_cure(action) {
    set_cure(String(action['color']), action['cured']);
    for ( var i=0; i<action['cards'].length; i++ ) {
        undo_discard(action['cards'][i], action, action['id']);
    }
    var city = document.getElementById('city-'+String(action['origin']));
    var dims = city.getBoundingClientRect();
    if ($(".city-"+String(action['origin'])+".cube-"+String(action['color'])).length !== action['cubes']) {
        var cube_counter = document.getElementById(String(action['color'])+"-cnt").getElementsByTagName('tspan')[0].textContent;
        var cubes_left = Number(cube_counter);
        add_cubes(String(action['origin']), dims, String(action['color']), action['rows'][action['color']], action['cubes']);
        stop_svg(action['origin']);
        if ( $.inArray(Number(action['origin']), action['at_risk']) !== -1 ) {
            pulse_svg(action['origin']);
        }

        cubes_left += action['cubes'];
        document.getElementById(String(action['color'])+"-cnt").getElementsByTagName('tspan')[0].textContent = String(cubes_left);
    }
    $('#logger').html('(UNDO) Removed '+COLORS[action['color']]+' cure. Returned cards to hand.');
    $('#make-cure').prop('disabled', false);
    $('#make-cure').attr('class', 'action');
}

function undo_take(action) {
    $('#'+action['giver']+'-card-'+action['card']).on('click', take_card).attr('class', 'card takeable').show(200);
    $('#card-'+action['card']).off().hide(200).attr('class', 'pl-card');
    set_cities(action['available']);
    $('#logger').html('(UNDO) Returned '+CARDS[Number(action['card'])].bold()+' to the '+ROLES[action['giver']].bold()+'.');
}

function undo_give(action) {
    $('#'+action['taker']+'-card-'+action['card']).off().attr('class', 'card').hide(200);
    $('#card-'+action['card']).off().on('click', give_card).attr('class', 'pl-card giveable').show(200);
    set_cities(action['available']);
    $('#logger').html('(UNDO) Returned '+CARDS[Number(action['card'])].bold()+' to the '+ROLES[action['giver']]+'.');
}

function undo_rp(action) {
    $('#infect-discard-'+String(action['deleted'])).off().attr('class', 'infect-discard');
    $('#logger').html('(UNDO) Returned '+CARDS[action['deleted']].bold()+' from the graveyard.')
    undo_discard('52', action, action['owner']);
}

function undo_oqn(action) {
    $('#logger').html('(UNDO) ');
    undo_discard('51', action, action['owner']);
}

function undo_store(action) {
    $('#logger').html('(UNDO) Return '+CARDS[action['card']]+' to the discard pile.');
    $('#cp-store').off().hide();
    $('#pl-discard-'+String(action['card'])).off().on('click', store_on_cp).addClass('pl-discard takeable').show();
}

function undo_trash(action) {
    var trash = action['trash'];
    if (typeof trash !== 'undefined') {
        if (trash['act'] === 'airlift') {
            undo_discard(trash['cards'], action, trash['owner']);
            undo_move(trash);
        } else if (trash['act'] === 'rp') {
            undo_rp(trash);
        } else if (trash['act'] === 'gg') {
            undo_station(trash);
        } else if (trash['act'] === 'oqn') {
            undo_oqn(trash);
        } else if (trash['act'] === 'trash') {
            undo_discard(trash['cards'], action, trash['owner']);
            set_cities(trash['available']);
        }
    }
}

File: app/views/undo.py
from app import app
from ..src.game import Game
from ..src.constants import *
from flask import render_template, flash, redirect, url_for, request, jsonify, \
                    session, Blueprint, g
import json
import random
import pickle
from copy import copy
import sys

undo = Blueprint('undo', __name__)

@undo.route('/_undo')
def undo_action():
    game = pickle.loads(session['game'])
    actions = session['actions']
    player = game.players[game.active]
    action = actions.pop(-1)
    data = action

    game.phase -= 1
    if 'trash' in data:
        trash = data['trash']

    undo_trash(data, game, player)
    if action['act'] == "drive" or action['act'] == "shuttle" or action['act'] == "dispatch":
        undo_move(data, player, game)
    elif action['act'] == "charter" or action['act'] == "fly" or action['act'] == "airlift":
        for p in game.players:
            if p.get_id() == data['owner']:
                owner = p
        undo_move(data, player, game)
        undo_discard(int(data['cards']), owner, game)
        if action['act'] == 'airlift':
            game.phase += 1
    elif action['act'] == "station-fly":
        undo_move(data, player, game)
        undo_discard(int(data['cards']), player, game)
        player.has_stationed = False
    elif action['act'] == "build" or action['act'] == "gg":
        undo_station(data, player, game)
        if action['act'] == "gg":
            game.phase += 1
    elif action['act'] == "treat":
        undo_treatment(data, game)
    elif action['act'] == "cure":
        undo_cure(data, player, game)
    elif action['act'] == "take":
        undo_take(data, game)
    elif action['act'] == "give":
        undo_give(data, game)
    elif action['act'] == "rp":
        undo_rp(data, game)
        game.phase += 1
    elif action['act'] == "oqn":
        undo_oqn(data, game)
        game.phase += 1
    elif action['act'] == "store":
        undo_store(data, game)

    available, dispatch, origin, player_id = game.set_available(0)
    can_take, can_give, team_hands = game.set_share()
    can_build = player.can_build(player.get_position(), game.research_stations)
    can_cure = player.can_cure(game.research_stations)

    session['actions'] = actions
    session['game'] = pickle.dumps(game)
    return jsonify( result=action,
                    available=available,
                    can_give=can_give,
                    can_take=can_take,
                    hand=player.hand,
                    team_hands=team_hands,
                    can_build=can_build,
                    can_cure=can_cure )

def undo_move(data, player, game):
    if player.get_id() != data['mover']:
        for p in game.players:
            if p.get_id() == data['mover']:
                p.position = data['origin']
                if data['mover'] == 'qs':
                    game.quarantined = game.board.get_neighbors(p.get_position())
                    game.quarantined.append(p.get_position())
    else:
        player.position = data['origin']
        if data['id'] == 'qs':
            game.quarantined = game.board.get_neighbors(player.get_position())
            game.quarantined.append(player.get_position())

    for color in COLORS:
        number = data['cubes'][color]
        if number != game.cubes[data['destination']][color]:
            game.cubes[data['destination']][color] = number
            game.cubes_left[color] -= number
            game.risk(data['destination'])
    game.board.rows[data['destination']] = data['rows']

def undo_discard(card, player, game):
    if card in game.player_cards.discard:
        game.player_cards.remove_from_discard(card)
        player.add_card(card)
    elif card in game.player_cards.graveyard:
        game.player_cards.remove_from_graveyard(card)
        player.event = card

def undo_station(data, player, game):
    game.research_stations.remove(int(data['origin']))
    if data['removed'] != '-1':
        game.research_stations.append(int(data['removed']))
    if data['cards'] != -1:
        for p in game.players:
            if p.get_id() == data['owner']:
                owner = p
        undo_discard(int(data['cards']), owner, game)

def undo_treatment(data, game):
    game.cubes_left[int(data['color'])] -= data['removed']
    game.cubes[int(data['origin'])][int(data['color'])] = data['cubes']
    game.board.rows[int(data['origin'])] = data['rows']
    game.risk(int(data['origin']))

def undo_cure(data, player, game):
    for card in data['cards']:
        game.player_cards.remove_from_discard(int(card))
        player.add_card(int(card))
    game.cures[data['color']] = LIVE
    game.cubes_left[data['color']] -= data['cubes']
    if data['medic_pos'] != -1:
        game.cubes[medic_pos][data['color']] = data['cubes']
        game.board.rows[medic_pos] = data['rows']
        game.risk(medic_pos)

def undo_take(data, game):
    for p in game.players:
        if p.get_id() == data['taker']:
            taker = p
        elif p.get_id() == data['giver']:
            giver = p
    giver.take_card(int(data['card']), taker)

def undo_give(data, game):
    print(data['taker'], data['giver'])
    for p in game.players:
        if p.get_id() == data['taker']:
            taker = p
        elif p.get_id() == data['giver']:
            giver = p
    taker.give_card(int(data['card']), giver)

def undo_rp(data, game):
    game.infect_cards.graveyard.remove(data['deleted'])
    game.infect_cards.add_to_discard(data['deleted'])
    for p in game.players:
        if p.get_id() == data['owner']:
            owner = p
    undo_discard(RP, owner, game)

def undo_oqn(data, game):
    game.oqn = False
    for p in game.players:
        if p.get_id() == data['owner']:
            owner = p
    undo_discard(OQN, owner, game)

def undo_store(data, game):
    player = game.players[game.active]
    player.event = None
    game.player_cards.add_to_discard(data['card'])

def undo_trash(data, game, player):
    if 'trash' in data:
        trash = data['trash']
        if 'owner' in trash:
            for p in game.players:
                if p.get_id() == trash['owner']:
                    owner = p
        if trash['act'] == 'airlift':
            undo_move(trash, player, game)
            undo_discard(AIRLIFT, owner, game)
        elif trash['act'] == 'gg':
            undo_station(trash, player, game)
        elif trash['act'] == 'rp':
            undo_rp(trash, game)
        elif trash['act'] == 'oqn':
            undo_oqn(trash, game)
        else:
            undo_discard(int(trash['cards']), owner, game)
